codestyle/styles.md

```diff
@@ -0,0 +1,672 @@
+}
+```
+
+**Empty rulesets**
+
+Empty rulesets are not allowed. A ruleset is considered as empty if it doesn't
+contain at least one styling rule. These record will be wrong:
+
+```CSS
+.block {}
+
+.comment-inside {
+    /*
+        Still empty because comments are not the styling rules
+    */
+}
+```
+
+The only exception is if a block doesn't have any styling rules, but its
+elements have you can keep the block ruleset empty:
+
+```CSS
+.block {}
+
+    .block__element1 {
+        background: #F00;
+    }
+
+    .block__element2 {
+        background: #0F0;
+    }
+```
+
+**Indenting rulesets**
+
+* one (1) empty line between rulesets;
+* indent rulesets to signal their relation to each other in HTML;
+
+Example:
+
+```CSS
+.block {
+    …
+}
+
+    .block__element {
+        …
+    }
+
+        .block__inner-element {
+            …
+        }
+```
+
+* `::before` and `::after` pseudo-elements must be treated as inner elements and
+indented accordingly:
+
+```CSS
+.block {
+    …
+}
+
+    .block::before {
+        ...
+    }
+
+    .block::after {
+        ...
+    }
+```
+
+
+
+<a id="selectors"></a>
+## Selectors
+
+* keep selectors as short as possible, in order to keep specificity down and
+  performance up;
+
+**Universal selector (`*`) is forbidden**
+
+```CSS
+/* Wrong: */
+
+* {
+    box-sizing: border-box;
+}
+```
+
+*Reasons:*
+* with high probability it will change rules that you were not going
+  to change;
+* because browsers read selectors right-to-left, the rightmost selector is often
+  critical in defining a selector’s performance;
+
+**No selectors by ID**
+
+```CSS
+/* Wrong: */
+
+#main {
+    color: #F00;
+}
+```
+
+*Reasons:*
+* selectors by id give too much weight for the rules under it;
+* it is impossible to use such blocks on the page more than one times;
+
+**No selectors by element**
+
+```CSS
+/* Wrong: */
+
+p {
+    margin: 1em 0;
+}
+
+.block img {
+    border: 1px solid #000;
+}
+```
+*Reasons:*
+* because with high probability the rules will be applied to elements you were
+  not going apply to;
+* inability to replace one tag on another without changing styles;
+
+*Exceptions:*
+* allowed to use in the reset/normalize styles files, which set basic
+  styles for the whole project;
+* allowed to use for styling selectors related to tables;
+
+**No selectors by attribute**
+
+```CSS
+/* Wrong: */
+
+.input[type="text"] {
+    font-size: 1em;
+}
+
+.input[disabled] {
+    opacity: 0.5;
+}
+
+
+/* Right: */
+
+.input_type_text {
+    font-size: 1em;
+}
+
+.input_state_disabled {
+    opacity: 0.5;
+}
+```
+
+*Reasons:*
+* the attribute selector has the same specificity as the class selector which
+  leads to an increase in the specificity and complexity of the redefinition of
+  the rules;
+
+*Exceptions:*
+* in case we need to style embedded a third-party widgets we have no ability to
+  edit its markup we can use this approach:
+
+```CSS
+[id="third-party-widget"] {}
+```
+
+
+**No cascade selectors**
```

Comment:
AFAIK, BEM doesn't forbid using cascade. If we forbid cascade __completely__ we will have to use mixes in case to _contextify_ block — and sometimes it's just overhead